src/goto-cc/linker_script_merge.h

```diff
@@ -0,0 +1,201 @@
+/// \file linker_script_merge.h
+/// \brief Merge linker script-defined symbols into a goto-program
+/// \author Kareem Khazem <karkhaz@karkhaz.com>
+
+#ifndef CPROVER_GOTO_CC_LINKER_SCRIPT_MERGE_H
+#define CPROVER_GOTO_CC_LINKER_SCRIPT_MERGE_H
+
+#include <functional>
+
+#include <util/cout_message.h>
+#include <util/json.h>
+
+#include "compile.h"
+#include "gcc_cmdline.h"
+
+/// \brief Patterns of expressions that should be replaced
+///
+/// Each instance of this class describes an expression 'shape' that should be
+/// replaced with a different expression. The intention is that if some
+/// expression matches the pattern described by this replacement_predicatet
+/// (i.e. replacement_predicatet::match() returns true), then that expression
+/// should be replaced by a new expression.
+class replacement_predicatet
+{
+public:
+  replacement_predicatet(
+      const std::string &description,
+      const std::function<const symbol_exprt(const exprt&)> inner_symbol,
+      const std::function<bool(const exprt&)> match)
+    : description(description),
+      inner_symbol(inner_symbol),
+      match(match)
+  {}
+
+  /// \brief a textual description of the expression that we're trying to match
+  const std::string description;
+
+  /// \brief Return the underlying symbol of the matched expression
+  /// \pre replacement_predicatet::match() returned true
+  const std::function<const symbol_exprt(const exprt&)> inner_symbol;
+
+  /// \brief Predicate: does the given expression match an interesting pattern?
+  ///
+  /// If this function returns true, the entire expression should be replaced by
+  /// a pointer whose underlying symbol is the symbol returned by
+  /// replacement_predicatet::inner_symbol().
+  const std::function<bool(const exprt&)> match;
+};
+
+/// \brief Synthesise definitions of symbols that are defined in linker scripts
+class linker_script_merget:public messaget
+{
+public:
+  /// \brief Add values of linkerscript-defined symbols to the goto-binary
+  /// \pre There is a single output file in each of `elf_binaries` and
+  ///      `goto_binaries`, and the codebase is being linked with a custom
+  ///      linker script passed to the compiler with the `-T` option.
+  /// \post The `__CPROVER_initialize` function contains synthesized definitions
+  ///       for all symbols that are declared in the C codebase and defined in
+  ///       the linker script.
+  /// \post All uses of these symbols throughout the code base are re-typed to
+  ///       match the type of the synthesized definitions.
+  /// \post The `__CPROVER_initialize` function contains one
+  ///       `__CPROVER_allocated_memory` annotation for each object file section
+  ///       that is specified in the linker script.
+  int operator()();
```

Comment:
Don't understand why this is a function-call-operator and not a method with a nice readable (self-documenting) name.